### PR TITLE
General Detction and Method Settings

### PR DESCRIPTION
**Pull request by andreitudorp**
#### API v3

!!! The UUID is now transmitted through a cookie with id `uuid` !!!

Added `MAX_CONTENT_LENGTH` to limit the size of the payloads that can be sent and processed by the API. (MAX. 10KiB)

New calls:
- ping [GET]
- settings [GET]
- settings [POST]
- TODO (in another PR): uuid [GET]

#### Detection Settings
Implemented the detection settings system which acts per user(uuid). The settings are stored in a table (in the settings database). There are some checks to make sure that the general structure of the JSON settings is correct. Multiple checks could be implemented in the future.

### Fixes
- naming inconsistencies
- formatting

### CODE REVIEW

File: methods/dst.py [hunk 1]
```diff
@@ -192,10 +207,14 @@ async def check_search_results(url_registered_domain, found_domains, worker: Thr
         future_group: FutureGroup = worker.new_future_group()
 
         for domain in found_domains:
-            future_group.schedule((url_registered_domain, domain), (lambda url_registered_domain, domain: check_url(url_registered_domain, domain)))
+            future_group.schedule(
+                (url_registered_domain, domain),
+                (lambda url_registered_domain, domain: check_url(url_registered_domain, domain)),
```
Comment (TPGamesNL) [inline methods/dst.py:212]:
```suggestion
                check_url,
```

File: utils/settings.py
Comment (TPGamesNL) [inline utils/settings.py:1]: why is this here?

File: utils/sessions.py [hunk 2]
```diff
@@ -3,6 +3,10 @@
 
 import tldextract
 
+from utils.logging import main_logger
+
+logger = main_logger.getChild("utils.sessions")
```
Comment (TPGamesNL) [inline utils/sessions.py:8]: is this logger used?

File: settings/__init__.py [hunk 3]
```diff
@@ -0,0 +1,8 @@
+from abc import ABC, abstractmethod
+
+
+class Settings(ABC):
```
Comment (TPGamesNL) [inline settings/__init__.py:4]: docs

File: settings/dst.py [hunk 4]
```diff
@@ -0,0 +1,16 @@
+from settings import Settings
+
+
+class DSTSettings(Settings):
+    # Which logo finder to use, 0 for `reverse_logo_region_search`, 1 for `vision_logo_detection`
+    LOGO_FINDERS: dict[str:int] = {"homebrew": 0, "google_vision": 1}
```
Comment (TPGamesNL) [inline settings/dst.py:6]: keys dont match comment

File: settings/storage.py [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+import sqlite3
+from datetime import datetime
+
+from settings.detection import DetectionSettings
+from utils.logging import main_logger
+
+logger = main_logger.getChild("settings.storage")
+
+
+class SettingsStorage:
```
Comment (TPGamesNL) [inline settings/storage.py:11]: Why do we open and close the db each time instead of keeping it open?

File: settings/storage.py [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+    def _setup_storage(self, storage_conn):
+        sql_q_db = """
+            CREATE TABLE IF NOT EXISTS "settings" (
+                "uuid"	string,
+                "settings" string,
+                "timestamp" string
+            );"""
+
+        storage_conn.execute(sql_q_db)
+        storage_conn.commit()
```
Comment (TPGamesNL) [inline settings/storage.py:28]: Doesnt it autocommit?

File: settings/storage.py (hunk 5 above)
Comment (TPGamesNL) [reply settings/storage.py:11]: also docs

File: api_versions/v3.py [hunk 7]
```diff
@@ -0,0 +1,85 @@
...
+# The storage interface for the sessions
+session_storage = detection.session_storage
+
+# The storage interface for the settings per user
+settings_storage = detection.settings_storage
+
+
+@v3.route("/ping", methods=["GET"])
+def ping():
+    return ("", 200)
```
Comment (TPGamesNL) [inline api_versions/v3.py:23]:
```suggestion
    return ""
```

Comment (TPGamesNL) [reply api_versions/v3.py:23]: same for below

File: settings/detection.py [hunk 8]
```diff
@@ -0,0 +1,64 @@
+from registry import DECISION_STRATEGIES, DETECTION_METHODS, DETECTION_METHODS_SETTINGS
+from settings import Settings
+from utils.logging import main_logger
+
+logger = main_logger.getChild("settings.detection")
+
+
+class DetectionSettings(Settings):
```
Comment (TPGamesNL) [inline settings/detection.py:8]: docs

File: api.py [hunk 9]
```diff
@@ -9,19 +9,24 @@
...
+from api_versions.v3 import v3
 from utils.logging import main_logger
 
 nest_asyncio.apply()
 
 # The storage interface for the sessions
 session_storage = detection.session_storage
 
+# The storage interface for the settings per user
+settings_storage = detection.settings_storage
```
[the remark this answers is not in the document]
Comment (andreitudorp) [reply api.py:21]: Good question

File: api.py [hunk 10]
```diff
@@ -9,19 +9,24 @@
...
+
 # Instantiate a logger for this HTTP API
 logger = main_logger.getChild("api")
 
 # Initiate Flask app
 app = Flask(__name__)
 app.config["DEBUG"] = False
+app.config["MAX_CONTENT_LENGTH"] = 10 * 1024  # 10KiB max payload size
```
[the remark this answers is not in the document]
Comment (andreitudorp) [reply api.py:29]: To prevent sending payloads that are too big and could crash or slow down the server. Not really necessary but useful.

File: api_versions/v2.py [hunk 11]
```diff
@@ -14,13 +14,17 @@
 # The storage interface for the sessions
 session_storage = detection.session_storage
 
+# The storage interface for the settings per user
+settings_storage = detection.settings_storage
```
[the remark this answers is not in the document]
Comment (andreitudorp) [reply api_versions/v2.py:18]: Forgot to remove it

File: api_versions/v2.py [hunk 12]
```diff
@@ -37,24 +41,10 @@ def get_state():
     return jsonify(result)
 
 
-@v2.route("/settings", methods=["POST"])
```
[the remark this answers is not in the document]
Comment (andreitudorp) [reply api_versions/v2.py:40]: Was not used, decided it should be fully part of the v3 API

File: methods/random.py [hunk 13]
```diff
@@ -10,10 +15,18 @@ class Random(DetectionMethod):
...
 
-    async def run(self, *args, **kwargs) -> ResultType:
+    async def run(self, a, b, c, settings: RandomSettings) -> ResultType:
         """
         This method computes a random result.
         """
 
-        rand = random.randint(-1, 1)
-        return ResultType(rand)
+        with TimeIt("random detection method"):
```
[the remark this answers is not in the document]
Comment (andreitudorp) [reply methods/random.py:23]: Consistency with other methods? I will remove it otherwise.

[the remark this answers is not in the document]
Comment (TPGamesNL) [reply methods/random.py:23]: It's just for getting some performance output, I'd remove it

File: settings/storage.py (hunk 6 above)
Comment (andreitudorp) [reply settings/storage.py:28]: No, actually. I forgot to add it and was confused why nothing was saving.

File: settings/detection.py (hunk 8 above)
Comment (andreitudorp) [reply settings/detection.py:8]: No, actually. I forgot to add it and was confused why nothing was saving.

File: settings/storage.py (hunk 5 above)
Comment (andreitudorp) [reply settings/storage.py:11]: Will add a TODO

File: detection.py [hunk 14]
```diff
@@ -92,35 +73,49 @@ def test(
...
+            raise (f"The detection method {method} is not implemented.")
+
         logger.info(f"Started running method {method}")
+
         results.append(
-            asyncio.run(DETECTION_METHODS[method].run(data.url, data.screenshot_url, data.uuid, data.pagetitle))
+            asyncio.run(
+                DETECTION_METHODS[method].run(
+                    data.url, data.screenshot_url, data.pagetitle, settings.methods_settings[method]
+                )
```
[the remark this answers is not in the document]
Comment (andreitudorp) [reply detection.py:114]: We will leave the generics part for the "Data" PR
